Update menu swipe

featheredtoast opened the pull request:
Refactor the mobile swipe events in Discourse:

Sets the styles through css properties so that behavior can be overridden in stylesheets if needed. Refactor finish animation , and implement these as simple CSS animations through a temporary animate class over the course of the animation.

Prefer element moves using translations, rather than other css properties as translations and opacity triggers are less resource intensive. From testing on my android, this makes the pull out menus MUCH smoother.

I also made all menu animations adhere to reduced motion preferences for a11y improvements.

featheredtoast committed:
DEV: replace swipe events to use translate rather than left/right

translate is better for animations. also use native css animations for opening
and closing.

FEATURE: adhere to reduced motion preference

featheredtoast committed:
update swipe events for mobile timeline

featheredtoast committed:
refactor: simplify handle pan done logic on menus

featheredtoast committed:
a11y: respect prefers reduced motion on mobile timeline

featheredtoast committed:
FIX: animate opacity on closing menu swipes as well

pmusaraj reviewed (approved):
This looks great! Looking forward to trying this out on iOS. 

There is a fair bit of jQuery usage here though that we should replace by vanilla JS.

featheredtoast committed:
DEV: reduce jquery usage on pan events

eviltrout commented on app/assets/javascripts/discourse/app/components/site-header.js:
```diff
@@ -38,17 +36,34 @@ const SiteHeaderComponent = MountWidget.extend(
 
     _animateOpening($panel) {
-      $panel.css({ right: "", left: "" });
+      const $headerCloak = $(".header-cloak");
```
Maybe it's worth removing all the jQuery while you're in there?

featheredtoast replied:
sure, can do!

featheredtoast committed:
DEV: reduce jquery usage on header and menus

featheredtoast committed:
DEV: reduce jquery use in site-header

eviltrout reviewed (approved):
Oooooh yeah that's nice!

featheredtoast committed:
DEV: remove jquery use in pan-events

featheredtoast commented:
Still doing a bit more cleanup, will also attempt to look into touch tests too.

featheredtoast committed:
DEV: more refactoring of jquery out to vanilla

featheredtoast committed:
DEV: refactor animate closing

featheredtoast committed:
DEV: better jquery deprecations for body

featheredtoast committed:
DEV: reduce jquery usage in site-header

featheredtoast committed:
DEV: remove jquery usage

featheredtoast committed:
DEV: remove jquery useage in docking

featheredtoast committed:
DEV: add tests for menu swipe events

correctly handle touch targets.

test is run in 50% zoom which means offsets and x of touch events need to be halved

featheredtoast committed:
dynamically deal with zoom and non zoom test environments

pmusaraj commented on app/assets/javascripts/discourse/app/components/site-header.js:
```diff
@@ -394,27 +392,26 @@ const SiteHeaderComponent = MountWidget.extend(
-            $panel.css({ top: menuTop + "px", [heightProp]: height });
-            $(".header-cloak").css({ top: menuTop + "px" });
+          if (
+            panel.style.top !== menuTop + "px" ||
```
```suggestion
            panel.style.top !== `${menuTop}px` ||
```

featheredtoast committed:
Update app/assets/javascripts/discourse/app/components/site-header.js

Co-authored-by: Penar Musaraj <pmusaraj@gmail.com>

featheredtoast committed:
DEV: touch tests

Refactor test window to use a transform rather than non-standard zoom property

featheredtoast commented:
OK, I've added tests as well as updated some code to allow those tests to run. The test environment was breaking a few things so I've cleaned some stuff up. The main points here are:

* Zoom methods on test windows -- through non-testem environments this was using a non-standard zoom property. Through testem environments, this was using a transform property. I've updated the non-testem to use more standard transforms for zooming.
* I've updated the detection for dropdown vs slide-in header modes. Mobile test detection on header should always be 'slide-in'. It takes the difference of header width vs document width, and in tests, the display window is always going to be tiny in comparison with the body.

featheredtoast commented:
@eviltrout when you have a moment, would you mind doing a quick sanity check for these? Reasons for the changes is in the comment above ☝️ 

https://github.com/discourse/discourse/pull/13277/files?file-filters%5B%5D=.js#diff-4ce6a279eb43abbfe77be11767e60a0561926e29dbe440d776886b5c7f4ea01b

https://github.com/discourse/discourse/pull/13277/files#diff-c5865c5d26ff1eef7ee0e664a9eec8ca10c045c828aa90a5d47bc7def5381dadL324

eviltrout reviewed (approved):
Looks good to me!
